### PR TITLE
Fix #12839: Text truncation in Win32 help dialog

### LINKED ISSUE #12839: [Bug]: Windows help dialog box truncates help text due to too small statically sized conversion buffers
### Version of OpenTTD

master, Windows

### Expected result

Running `openttd.exe -h` shows the help text in a dialog box

### Actual result

The statically sized buffers for line ending conversion and char to wchar_t conversion in HelpDialogFunc are too small (when a modest number of AIs and GSs are installed).
Screenshot_2024-07-03_00-46-51

(The GS section of the output is missing entirely in the above screenshot).

### Steps to reproduce

Run  `openttd.exe -h` on Windows

### PR DESCRIPTION
**Pull request by JGRennison**
## Motivation / Problem

#12839

## Description

Replace statically sized buffers with dynamically sized buffers allocated using std::make_unique<[]>.

## Limitations

N/A

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/os/windows/win32.cpp [hunk 1]
```diff
@@ -158,31 +158,34 @@ void CreateConsole()
...
 						break;
 					}
 				}
-				*q++ = *p++;
+				*q++ = c;
 			}
-			*q = '\0';
+			*q++ = '\0';
 			/* We need to put the text in a separate buffer because the default
 			 * buffer in OTTD2FS might not be large enough (512 chars). */
```
Comment (LordAro) [inline src/os/windows/win32.cpp:185]: is this comment still accurate?

File: src/os/windows/win32.cpp [hunk 2]
```diff
@@ -158,31 +158,34 @@ void CreateConsole()
...
-			wchar_t help_msg_buf[8192];
-			SetDlgItemText(wnd, 11, convert_to_fs(help_msg, help_msg_buf, lengthof(help_msg_buf)));
+			const size_t help_msg_buf_size = ((q - help_msg.get()) * 3) / 2;
```
Comment (LordAro) [inline src/os/windows/win32.cpp:186]: i feel like this value might need explaining

File: src/os/windows/win32.cpp [hunk 3]
```diff
@@ -158,31 +158,34 @@ void CreateConsole()
...
 static INT_PTR CALLBACK HelpDialogFunc(HWND wnd, UINT msg, WPARAM wParam, LPARAM)
 {
 	switch (msg) {
 		case WM_INITDIALOG: {
-			char help_msg[8192];
-			const char *p = _help_msg;
-			char *q = help_msg;
-			while (q != lastof(help_msg) && *p != '\0') {
-				if (*p == '\n') {
+			const size_t help_msg_size = 1 + _help_msg.size() + std::count(_help_msg.begin(), _help_msg.end(), '\n');
```
Comment (LordAro) [inline src/os/windows/win32.cpp:168]: a brief comment along the lines of "account for adding \r to \n" would help

File: src/os/windows/win32.cpp (hunk 2 above)
Comment (PeterN) [reply src/os/windows/win32.cpp:186]: This is also obsolete. `OTTD2FS` doesn't have a length limit these days, it works with std::strings and uses as much as necessary.

i.e. `std::wstring help_msg_buf = OTTD2FS(help_msg);` should be all that's needed.

**Comment by PeterN** [comment]
Something like this might be better. No fixed length buffers. Untested.

https://github.com/OpenTTD/OpenTTD/compare/master...PeterN:OpenTTD:win32-help

**Comment by JGRennison** [comment]
> Something like this might be better. No fixed length buffers. Untested.
> 
> [master...PeterN:OpenTTD:win32-help](https://github.com/OpenTTD/OpenTTD/compare/master...PeterN:OpenTTD:win32-help)

I'd be happy to defer to that, instead of this PR.

**Comment by JGRennison** [comment]
Replaced by #12844

**Pull request closed without merging**